quic-rpc-derive/src/lib.rs

```diff
@@ -0,0 +1,230 @@
+use proc_macro::TokenStream;
+use proc_macro2::{Span, TokenStream as TokenStream2};
+use quote::{quote, ToTokens};
+use std::collections::{BTreeMap, HashSet};
+use syn::{
+    parse::{Parse, ParseStream},
+    parse_macro_input,
+    spanned::Spanned,
+    Data, DeriveInput, Fields, Ident, Token, Type,
+};
+
+const SERVER_STREAMING: &str = "server_streaming";
+const CLIENT_STREAMING: &str = "client_streaming";
+const BIDI_STREAMING: &str = "bidi_streaming";
+const RPC: &str = "rpc";
+const TRY_SERVER_STREAMING: &str = "try_server_streaming";
+const IDENTS: [&str; 5] = [
+    SERVER_STREAMING,
+    CLIENT_STREAMING,
+    BIDI_STREAMING,
+    RPC,
+    TRY_SERVER_STREAMING,
+];
+
+fn generate_rpc_impls(
+    pat: &str,
+    mut args: RpcArgs,
+    service_name: &Ident,
+    request_type: &Type,
+    attr_span: Span,
+) -> syn::Result<TokenStream2> {
+    let res = match pat {
+        RPC => {
+            let response = args.get("response", pat, attr_span)?;
+            quote! {
+                impl ::quic_rpc::pattern::rpc::RpcMsg<#service_name> for #request_type {
+                    type Response = #response;
+                }
+            }
+        }
+        SERVER_STREAMING => {
+            let response = args.get("response", pat, attr_span)?;
+            quote! {
+                impl ::quic_rpc::message::Msg<#service_name> for #request_type {
+                    type Pattern = ::quic_rpc::pattern::server_streaming::ServerStreaming;
+                }
+                impl ::quic_rpc::pattern::server_streaming::ServerStreamingMsg<#service_name> for #request_type {
+                    type Response = #response;
+                }
+            }
+        }
+        BIDI_STREAMING => {
+            let update = args.get("update", pat, attr_span)?;
```

Comment:
should `update` and `response` be constants as well?

Reply:
Not sure. They are only used once, in exactly one position, per rpc pattern. Sure, "update" and "result" are used multiple times in multiple patterns, but that's not really the same thing, just happens to have the same name.

The pattern names are used in 2 locations, in retain and in the switch.